userprofile/permissions.py

```diff
@@ -27,6 +28,13 @@ def filter_queryset(self, request, queryset, view):
         return queryset
 
 
+class IsTeacherOrAdminOrSelf(IsAdminOrUserObjIsSelf):
+    def is_super(self, user):
+        every_course = CourseInstance.objects.all()
+        is_teacher = any([course.is_teacher(user) for course in every_course])
```

Comment:
Some course permissions avoid looping over all courses in order to check the teachers by using custom attributes of the `view` variable in the `has_object_permission` method. For example, `view.is_teacher`:

* https://github.com/apluslms/a-plus/blob/6da3b36fbc583e40057c8119b25fc1a8997721f8/course/permissions.py#L150
* https://github.com/apluslms/a-plus/blob/6da3b36fbc583e40057c8119b25fc1a8997721f8/course/viewbase.py#L62 (this sets the `is_teacher` attribute to the view instance if the view class inherits `CourseInstanceBaseMixin`)
* https://github.com/apluslms/a-plus/blob/6da3b36fbc583e40057c8119b25fc1a8997721f8/course/api/views.py#L123 (example viewset in the API that inherits `CourseInstanceBaseMixin` via `CourseResourceMixin`)
* https://github.com/apluslms/a-plus/blob/6da3b36fbc583e40057c8119b25fc1a8997721f8/course/api/mixins.py#L13

There is still the problem that `UserViewSet` is not tied to any single course and of course you can't make it inherit `CourseInstanceBaseMixin`.